storage/shard.go

```diff
@@ -1542,6 +1568,68 @@ func (s *dbShard) Flush(
 	return s.markFlushStateSuccessOrError(blockStart, multiErr.FinalError())
 }
 
+func (s *dbShard) Snapshot(
+	blockStart time.Time,
+	snapshotTime time.Time,
+	flush persist.Flush,
+) error {
+	// We don't snapshot data when the shard is still bootstrapping
+	s.RLock()
+	if s.bs != bootstrapped {
+		s.RUnlock()
+		return errShardNotBootstrappedToSnapshot
+	}
+	s.RUnlock()
+
+	var multiErr xerrors.MultiError
+
+	s.markIsSnapshotting()
+	defer func() {
+		s.markDoneSnapshotting(multiErr.Empty(), snapshotTime)
+	}()
+
+	prepareOpts := persist.PrepareOptions{
+		NamespaceMetadata: s.namespace,
+		Shard:             s.ID(),
+		BlockStart:        blockStart,
+		SnapshotTime:      snapshotTime,
+		FilesetType:       persist.FilesetSnapshotType,
+	}
+	prepared, err := flush.Prepare(prepareOpts)
```

Comment:
don't you need to check `err != nil` before proceeding here?

Reply:
fixed for snapshot and flush method